src/components/WalletConnect/NanoContract/NewNanoContractTransactionRequest.js

```diff
@@ -0,0 +1,313 @@
+  const blueprintInfo = useSelector((state) => state.nanoContract.blueprint[nc.blueprintId]);
+
+  const [showSelectAddressModal, setShowSelectAddressModal] = useState(false);
+  const [showDeclineModal, setShowDeclineModal] = useState(false);
+  // If nano-contract's method is 'initialize' then the expression
+  // should be resolved to firstAddress value by default.
+  // However, the wallet can have trouble to
+  const [ncAddress, setNcAddress] = useState(registeredNc?.address || firstAddress.address);
+  const ncToAccept = useMemo(() => ({
+    ...nc,
+    caller: ncAddress,
+  }), [ncAddress])
+
+  // Controle SelectAddressModal
+  const toggleSelectAddressModal = () => setShowSelectAddressModal(!showSelectAddressModal);
+  const handleAddressSelection = (newAddress) => {
+    setNcAddress(newAddress);
+    toggleSelectAddressModal();
+  };
+
+  // Accepts the Nano Contract data preseted.
+  const onAcceptTransaction = () => {
+    // Update the caller with the address selected by the user.
+    const acceptedNc = { ...nc, caller: ncAddress };
+    // Signal the user has accepted the current request and pass the accepted data.
+    dispatch(walletConnectAccept(acceptedNc));
+  };
+
+  const onDeclineTransaction = () => {
+    setShowDeclineModal(true);
+  };
+  const onDeclineConfirmation = () => {
+    setShowDeclineModal(false);
+    dispatch(walletWalletReject());
+    navigation.goBack();
+    // TODO: It is not being dimissed automatically
+  };
+  const onDismissDeclineModal = () => {
+    setShowDeclineModal(false);
+  };
+
+  // Control which content to show, if the nano contract is not registered
+  // a feedback content should tell user the nano contract must be registered first
+  // and only let user decline the transaction to get out the page, otherwise interaction
+  // content is showed.
+  const notRegistered = ncToAccept.method !== 'initialize' && registeredNc == null;
+  const isRegistered = !notRegistered;
+
+  // This effect should run only once because firstAddress is kept on state when loaded
+  useEffect(() => {
+    if (firstAddress.address) {
+      setNcAddress(firstAddress.address);
+    }
+  }, [firstAddress]);
+
+  // This effect runs only once in the construct phase
+  useEffect(() => {
+    // Do nothing if nano contract is not registered and don't call initialize method.
+    if (notRegistered) return;
+
+    // Request blueprint info if not present to feed the components:
+    // - NanoContractExecInfo, and
+    // - NanoContractMethodArgs
+    if (!blueprintInfo) {
+      dispatch(nanoContractBlueprintInfoRequest(nc.blueprintId));
+    }
+
+    // Get tokens metadata
+    // actions can be null?
+    const tokensUid = nc.actions?.map((each) => each.token) || [];
+    const tokensMetadataToDownload = [];
+
+    tokensUid.forEach((uid) => {
+      if (uid !== DEFAULT_TOKEN.uid && !(uid in registeredTokensMetadata)) {
+        tokensMetadataToDownload.push(uid);
+      }
+    });
```

Comment:
Why do you need this? If I'm not wrong, this token metadata is used to get NFT and metadata information. Why do you need to get these from the actions tokens?

Reply:
You right! I refactored it to use the right API.